bftclient/src/bft_client.cpp

```diff
@@ -156,42 +204,88 @@ Reply Client::send(const MatchConfig& match_config,
     if (auto reply = wait()) {
       expected_commit_time_ms_.add(
           std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - start).count());
+      reply_certificates_.clear();
       return reply.value();
     }
     metrics_.retransmissions.Get().Inc();
   }
 
   expected_commit_time_ms_.add(request_config.timeout.count());
-  outstanding_request_ = std::nullopt;
+  reply_certificates_.clear();
   throw TimeoutException(request_config.sequence_number, request_config.correlation_id);
 }
 
+std::deque<Reply> Client::sendBatch(std::deque<WriteRequest>& write_requests, const std::string& cid) {
+  std::deque<Reply> replies;
+  std::chrono::milliseconds max_time_to_wait = 0s;
+  MatchConfig match_config = writeConfigToMatchConfig(write_requests.front().config);
+  auto batch_msg = initBatch(write_requests, cid, max_time_to_wait);
+  auto start = std::chrono::steady_clock::now();
+  auto end = start + max_time_to_wait;
+  while (std::chrono::steady_clock::now() < end && replies.size() != pending_requests_.size()) {
+    bft::client::Msg msg(batch_msg);  // create copy here due to the loop
```

Comment:
Why do we need to have a copy here?

Reply:
We have like retry mechanism, we are sending the copied message to the replicas then we are going into the wait() method and waiting for replies, if replies don't arrive (we got dynamic utility for replies timeouts) and we still have time we are sending again the same message.
since we are moving the message so we need to move copies.